src/library/Query.js

```diff
@@ -56,47 +56,36 @@ function limitQuery(httpMethod, deployment, apiKey, queryPath, limit, offset, pa
   const separationLimit = 30;
   const limitChecked = checkLimit(limit);
   const offsetChecked = checkOffset(offset);
-  if (limitChecked > separationLimit) {
-    const results = {
-      status: null,
-      message: null,
-    };
-    // MBQUERY, MBCUSTOMQUERY has rows in result but MBEVENTS has not
-    const hasRows = !/^events$/.test(queryPath);
-    const rows = [];
-
-    let offsetNext = offsetChecked;
-    while (true) {
-      // validate and normalize limit and offset
-      const queryOptions = buildQueryOptions(separationLimit, offsetNext, address);
-
-      // query
-      const queryResult = query(httpMethod, deployment, apiKey, queryPath + queryOptions, payload);
-      results.status = queryResult.status;
-      results.message = queryResult.message;
-
-      const queryRows = hasRows ? queryResult.result.rows : queryResult.result;
-      const preLength = rows.length + queryRows.length;
-      const currentLength = queryRows.length;
-      if (currentLength < 1) {
-        break;
-      } else if (preLength > limitChecked) {
-        // Get rid of the overflowed if actual data is more than limitChecked(limit)
-        const cutOff = currentLength - (preLength - limitChecked);
-        rows.push(...queryRows.slice(0, cutOff));
-        break;
-      } else {
-        rows.push(...queryRows);
-      }
-
-      offsetNext += separationLimit;
-    }
-
-    results.result = hasRows ? { rows } : rows;
-
-    return results;
+  const results = {
+    status: null,
+    message: null,
+  };
+  // MBQUERY, MBCUSTOMQUERY has rows in result but MBEVENTS has not
+  const hasRows = !/^events$/.test(queryPath);
+  let rows = [];
+  let queryRows = [];
+
+  let offsetNext = offsetChecked;
+  do {
+    // validate and normalize limit and offset
+    const queryOptions = buildQueryOptions(separationLimit, offsetNext, address);
+
+    // query
+    const queryResult = query(httpMethod, deployment, apiKey, queryPath + queryOptions, payload);
+    queryRows = hasRows ? queryResult.result.rows : queryResult.result;
+    rows.push(...queryRows);
```

Comment:
I don't want to get too fancy, but could we do something like:

```js
// either add the entire array of "queryRows" to "rows", or only up to the total limit
// works because if "end" is greater than the length of the array, it only extracts
// up to the end of the array
// see: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice
rows.push(...queryRows.slice(0, limitChecked - rows.length));
```

Reply:
limitChecked will be `Infinity` when a limit is -1, for this I added `queryRows.length > 0` in do {} while() and finalize rows after a loop. But I can change some and get slice back into a loop to avoid slice for a giant array.

Reply:
Ah true! Maybe leave as-is then. We probably should leave the optimizing to the compiler at this point, you cancan disregard my suggestion!